### DISCUSSION TITLE
How to register a new lexer with TextAdept?

### DISCUSSION THREAD

**Discussion by nsgnkhibdk2cls0f** [orbitalquark/textadept#312] [Q&A] [answered]
The lexer is not automatically selected when I opened the file. I have to tell TextAdept to use it each time I opened the file with this extension.

**Comment by orbitalquark** [comment]
It depends which version of Textadept you are using. If you're using 11.4 or lower you would add to the `textadept.file_types.extensions` table. If you're using 11.5 alpha or the nightly version, you would add to `lexer.detect_extensions`: https://orbitalquark.github.io/textadept/api.html#lexer.detect_extensions

    -- Textadept 11.4 or lower.
    textadept.file_types.extensions.my_ext = 'my_lexer'
    -- Textadept 11.5 alpha or higher.
    lexer.detect_extensions.my_ext = 'my_lexer'

Technically there is a compatibility layer in 11.5 alpha and the nightlies that allow you to use 11.4's API. This will not be there forever though.

**Reply by nsgnkhibdk2cls0f** [reply]
Where to put `lexer.detect_extensions`?

**Reply by nsgnkhibdk2cls0f** [reply]
In the lexer file or init.lua?

**Reply by nsgnkhibdk2cls0f** [reply]
@orbitalquark Isn't it better to tell me how to use it rather than throwing the api index into my face and gone my friend? I'm not a Lua guy I have to Google Lua syntax all the time. `lexer.detect_extensions` is said to be a map. Lua only has table but no such thing as a map. Either way both `lexer.detect_extensions` alone or `lexer.detect_extensions = true` or `lexer.detect_extensions = {foo = 'bar'}` do not work. It's only when I found your very own [.textadept directory](https://github.com/orbitalquark/.textadept) did I knew to put it on init.lua and inspired me to try this: `lexer.detect_extensions.foo = 'bar'` but it still not work my friend.

**Reply by nsgnkhibdk2cls0f** [reply]
I'm using your latest nightly. Please give me a working answer. Thank you.

**Reply by nsgnkhibdk2cls0f** [reply]
Your manual doesn't provide more information than your answer:

https://orbitalquark.github.io/textadept/manual.html#working-with-files-and-projects

I will wait for your reply.

**Comment by orbitalquark** [comment, marked as answer]
I agree that reading the API documentation as a newcomer to Lua would be quite intimidating. I recommend reading the Textadept manual first: https://orbitalquark.github.io/textadept/manual.html. You'll find help regarding lexer detection here: https://orbitalquark.github.io/textadept/manual.html#language.

Your use of `lexer.detect_extensions.foo = 'bar'` should work, provided you put a working *bar.lua* lexer in either *~/.textadept/lexers/* or Textadept's *lexers/* directory.

**Reply by nsgnkhibdk2cls0f** [reply]
Yeah, I know why it didn't work now. foo is the lexer name and bar is the extension, after I swapped them it worked:

`lexer.detect_extensions.bar = 'foo'`

The luadoc example is confusing, as newcomers think the extension is lua and the lexer is luadoc when in fact it's the other way around.
